src/node_api.cc

```diff
@@ -3554,3 +3557,299 @@ napi_status napi_run_script(napi_env env,
+                ts_fn->thread_finalize_callback = thread_finalize_cb;
+                ts_fn->thread_finalize_data = thread_finalize_data;
+
+                *result = reinterpret_cast<napi_threadsafe_function>(ts_fn);
+                return napi_clear_last_error(env);
+              }
+              if (max_queue_size != 0) {
+                uv_cond_destroy(&ts_fn->cond);
+              }
+            }
+            v8impl::Reference::Delete(ts_fn->ref);
+          }
+          uv_mutex_destroy(&ts_fn->mutex);
+        }
+        uv_close(reinterpret_cast<uv_handle_t*>(&ts_fn->async),
+            [] (uv_handle_t* handle) -> void {
+              delete handle;
+            });
+
+        // Prevent the thread-safe function from being deleted here, because the
+        // uv_close() callback above will delete it.
+        ts_fn = nullptr;
+      }
+      delete ts_fn;
+    }
+  }
+  return napi_set_last_error(env, status);
+}
+
+NAPI_EXTERN napi_status
+napi_get_threadsafe_function_hint(napi_threadsafe_function ts_fn,
+                                  void** result) {
+  CHECK(ts_fn != nullptr);
+  CHECK(result != nullptr);
+
+  *result = ts_fn->hint;
+  return napi_ok;
+}
+
+NAPI_EXTERN napi_status
+napi_call_threadsafe_function(napi_threadsafe_function ts_fn,
+                              void* data,
+                              bool block_when_full) {
+  CHECK(ts_fn != nullptr);
+
+  napi_status status = napi_ok;
+
+  uv_mutex_lock(&ts_fn->mutex);
+  if (!ts_fn->is_closing) {
+    while (ts_fn->queue.size() >= ts_fn->max_queue_size &&
+        ts_fn->max_queue_size > 0) {
+      if (!block_when_full) {
+        status = napi_queue_full;
+        break;
+      }
+      uv_cond_wait(&ts_fn->cond, &ts_fn->mutex);
+    }
+    if (status == napi_ok) {
+      if (uv_async_send(&ts_fn->async) == 0) {
+        ts_fn->queue.push(data);
+      } else {
+        status = napi_generic_failure;
+      }
+    }
+  } else {
+    status = napi_closing;
+  }
+  uv_mutex_unlock(&ts_fn->mutex);
+
+  return status;
+}
+
+static napi_status
+napi_tick_threadsafe_function(napi_threadsafe_function ts_fn,
+                              bool unref,
+                              bool abort) {
+  CHECK(ts_fn != nullptr);
+  napi_status status = napi_ok;
+
+  uv_mutex_lock(&ts_fn->mutex);
+  if (ts_fn->refcount == 0) {
+    status = napi_invalid_arg;
+  } else {
+    ts_fn->refcount += (unref ? -1 : 1);
+  }
+
+  if (!ts_fn->is_closing) {
+    ts_fn->is_closing = (ts_fn->refcount == 0) || abort;
+  }
+
+  if (uv_async_send(&ts_fn->async) != 0) {
+    status = napi_generic_failure;
+  }
+  uv_mutex_unlock(&ts_fn->mutex);
+
+  return status;
+}
+
+NAPI_EXTERN napi_status
+napi_ref_threadsafe_function(napi_threadsafe_function ts_fn) {
+  return napi_tick_threadsafe_function(ts_fn, false, false);
+}
+
+NAPI_EXTERN napi_status
+napi_unref_threadsafe_function(napi_threadsafe_function ts_fn,
+                               bool abort) {
```

Comment:
We’ve been trying to avoid boolean arguments in our APIs if a custom 2-value `enum` would also do the trick, because that usually leads to clearer API usage